source/controllers/sudokuWorld.d

```diff
@@ -0,0 +1,161 @@
+
+import controllers.createMenuController;
+import controllers.sudokuController;
+import core.sudoku.grid;
+import core.sudoku.sudoku;
+import core.sudokuType;
+import ui.menus.createMenu;
+import ui.sudoku.gridUI;
+import ui.sudokuApp;
+
+
+public class SudokuWorld
+{
+	public this(SudokuApp app)
+	{
+		trace("Initializing SudokuWorld");
+		_app = app;
+		_createMenuController = new CreateMenuController(new CreateMenu(app.builder), this);
+		callbacks();
+	}
+
+
+	private void callbacks()
+	{
+		trace("Initializing SudokuWorld callbacks...");
+		app.setOnBtnCreateClicked(&onBtnCreateClicked);
+		app.setOnBtnExitClicked(&onBtnExitClicked);
+		app.setOnButtonPress(&onButtonPress);
+	}
+
+
+	/** This is called when `btnCreate` is clicked in the main menu
+	 *
+	 * Makes the create menu content `Stack` visible
+	 */
+	private void onBtnCreateClicked(Button)
+	{
+		app.showCreateMenu();
+	}
+
+
+	/** Exit the program
+	 */
+	private void onBtnExitClicked(Button)
+	{
+		app.quit();
+	}
+
+
+	// TODO: ui:sudokuWorld: implement Play Menu
+	private void onBtnPlayClicked(Button)
+	{
+		// do something
+	}
+
+
+	// TODO: ui:sudokuWorld: implement Settings Menu
+	private void onBtnSettingsClicked(Button)
+	{
+		// do something
+	}
+
+
+	private bool onButtonPress(GdkEventButton* ev, Widget widget)
+	{
+		if (ev.button != 1)
+			return false;
+
+
+		// clear focus in the active grid
+		GridUI gridUI;
+		if (GameState.Play in sudokuControllers)
+		{
+			gridUI = sudokuControllers[GameState.Play].gridUI;
+		}
+		else if (GameState.Solution in sudokuControllers)
+		{
+			gridUI = sudokuControllers[GameState.Solution].gridUI;
+			_createMenuController.setCellInfoText(0,0,0);
+		}
+		else if (GameState.Create in sudokuControllers)
+		{
+			gridUI = sudokuControllers[GameState.Create].gridUI;
+			_createMenuController.setCellInfoText(0,0,0);
+		}
+		gridUI.cleanFocus();
+		return false;
+	}
+
+
+
+// Getters/Setters
+
+	public void addSudokuController(SudokuController sudokuController)
+	{
+		sudokuControllers[sudokuController.gameState] = sudokuController;
+	}
+
+
+	public auto addSudokuController(GameState state, SudokuType type)
+	{
+		Grid grid = new Grid(Sudoku.dimension(type).expand);
+		grid.initialize(new int[][](grid.height, grid.width));
+		GridUI gridUI = new GridUI(type);
+		SudokuController sd = new SudokuController(grid, gridUI, state, this);
+		return sudokuControllers[state] = sd;
+	}
+
+
+	public auto addSudokuController(GameState state, SudokuController sudokuController)
+	{
+		SudokuType type = sudokuController.gridUI.type;
+		Grid grid = new Grid(Sudoku.dimension(type).expand);
+		int[][] digits = sudokuController.grid.toDigit();
+		grid.initialize(digits.dup);
+		GridUI gridUI = new GridUI(type);
+		SudokuController sd = new SudokuController(grid, gridUI, state, this);
+		return sudokuControllers[state] = sd;
+	}
+
+
+	public void deleteSudokuController(GameState state)
+	{
+		SudokuController* p = (state in sudokuControllers);
```

Comment:
Because inside a SudokuController lies a GridUI which is a Gtk Widget. The problem with this is, removing the controller from the associative array doesn't grant the immediate destruction of the object. This can cause conflict with Gtk because the Widget might be associated with another Widget, in this case, it is; the GridUI is contained in a Stack. There is no simple way of removing a Widget from another Widget, actualy I don't know any, the solution is destroying the object manually. Calling `destroy` will remove destroy the object and remove any association with another Widget. This way we can ensure `GridUI` is removed from the GtkStack.

Reply:
Oh, okay. In that case, maybe just a small comment here stating that helps, so later on this doubt doesn't come up again.